gfx/drivers/gl_core.c

```diff
@@ -0,0 +1,2001 @@
+      texture.format = GL_RGBA8;
+      texture.padded_width = gl->hw_render_max_width;
+      texture.padded_height = gl->hw_render_max_height;
+
+	  if (texture.width == 0)
+		  texture.width = 1;
+	  if (texture.height == 0)
+		  texture.height = 1;
+   }
+   else
+   {
+      texture.image = streamed->tex;
+      texture.format = gl->video_info.rgb32 ? GL_RGBA8 : GL_RGB565;
+      texture.padded_width = streamed->width;
+      texture.padded_height = streamed->height;
+   }
+   gl_core_filter_chain_set_input_texture(gl->filter_chain, &texture);
+   gl_core_filter_chain_build_offscreen_passes(gl->filter_chain, &gl->filter_chain_vp);
+
+   glBindFramebuffer(GL_FRAMEBUFFER, 0);
+   glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
+   glClear(GL_COLOR_BUFFER_BIT);
+   gl_core_filter_chain_build_viewport_pass(gl->filter_chain, &gl->filter_chain_vp,
+                                            gl->hw_render_bottom_left ? gl->mvp.data : gl->mvp_yflip.data);
+   gl_core_filter_chain_end_frame(gl->filter_chain);
+
+#if defined(HAVE_MENU)
+   if (gl->menu_texture_enable)
+   {
+      menu_driver_frame(video_info);
+      if (gl->menu_texture_enable && gl->menu_texture)
+         gl_core_draw_menu_texture(gl, video_info);
+   }
+   else if (video_info->statistics_show)
+   {
+      struct font_params *osd_params = (struct font_params*)
+         &video_info->osd_stat_params;
+
+      if (osd_params)
+      {
+         font_driver_render_msg(video_info, NULL, video_info->stat_text,
+               (const struct font_params*)&video_info->osd_stat_params);
+      }
+   }
+
+#ifdef HAVE_MENU_WIDGETS
+   menu_widgets_frame(video_info);
+#endif
+#endif
+
+#ifdef HAVE_OVERLAY
+   if (gl->overlay_enable)
+      gl_core_render_overlay(gl, video_info);
+#endif
+
+   if (!string_is_empty(msg))
+   {
+      //if (video_info->msg_bgcolor_enable)
+      //   gl_core_render_osd_background(gl, video_info, msg);
+      font_driver_render_msg(video_info, NULL, msg, NULL);
+   }
+
+   video_info->cb_update_window_title(
+         video_info->context_data, video_info);
+
+   if (gl->readback_buffer_screenshot)
+   {
+      /* For screenshots, just do the regular slow readback. */
+      gl_core_slow_readback(gl, gl->readback_buffer_screenshot);
+   }
+   else if (gl->pbo_readback_enable)
+   {
+#ifdef HAVE_MENU
+      /* Don't readback if we're in menu mode. */
+      if (!gl->menu_texture_enable)
+#endif
+         gl_core_pbo_async_readback(gl);
+   }
+
+   /* Disable BFI during fast forward, slow-motion,
+    * and pause to prevent flicker. */
+   if (
+         video_info->black_frame_insertion
+         && !video_info->input_driver_nonblock_state
+         && !video_info->runloop_is_slowmotion
+         && !video_info->runloop_is_paused)
+   {
+      video_info->cb_swap_buffers(video_info->context_data, video_info);
+      glClear(GL_COLOR_BUFFER_BIT);
+   }
+
+   video_info->cb_swap_buffers(video_info->context_data, video_info);
+
+   if (video_info->hard_sync &&
```

Comment:
just a heads up if this is done and there are comparisions drawn with the old GL driver, the old GL driver will be markedly faster because hard_sync will be in place while fast forwarding.

We noticed that fast forwarding was slower with GL2 when hard sync was enabled and we mitigated that effect (since it's pointless to hard sync while ff) with

```C
  /* check if we are fast forwarding or in menu, if we are ignore hard sync */
   if (  gl->have_sync
         && video_info->hard_sync
         && !video_info->input_driver_nonblock_state
         && !gl->menu_texture_enable)```

Reply:
guess I'll test instead of speculate 🙄

Reply:
hmmm it's already ~150fps faster in snes9x (super mario world) so I guess this is irrelevant

Reply:
We can do 
```cpp
/* check if we are fast forwarding or in menu, if we are ignore hard sync */
   if (  gl->vsync
         && video_info->hard_sync
         && !video_info->input_driver_nonblock_state
         && !gl->menu_texture_enable)
   {
      gl_core_fence_iterate(gl, video_info->hard_sync_frames);
   }
```
I see the max fps is 1400 vs 1100 before with hard sync enabled at 0 in snes9x when FFing.
The condition is fine right? I'm not disabling it completely or something?

Reply:
Yeah I tested the workaround but I saw no differences in my own testing, maybe I'm cpu bottlenecked :P

Reply:
The condition was flipped. Fixed now.